File: src/main/java/app/editor/components/ComponentButton.java
package app.editor.components;

import javafx.event.ActionEvent;
import javafx.event.EventHandler;
import javafx.scene.Cursor;
import javafx.scene.SnapshotParameters;
import javafx.scene.control.ContextMenu;
import javafx.scene.control.MenuItem;
import javafx.scene.control.SeparatorMenuItem;
import javafx.scene.image.Image;
import javafx.scene.image.ImageView;
import javafx.scene.input.*;
import javafx.scene.paint.Color;
import section.model.Component;
import sun.reflect.generics.reflectiveObjects.NotImplementedException;
import utils.OSValidator;

import java.util.ArrayList;
import java.util.List;

public class ComponentButton extends DragButton {
    private OnComponentActionListener onComponentActionListener;
    protected Component associatedComponent;
    private boolean isSelected = false;

    public ComponentButton(ComponentGrid componentGrid, Component associatedComponent) {
        super(componentGrid);
        this.associatedComponent = associatedComponent;

        // Add the style
        getStyleClass().add("component-btn");
        if (!associatedComponent.getItem().isValid()) {  // Invalid item rendering
            getStyleClass().add("invalid-btn");
        }

        // Set the button properties
        setMaxSize(Double.MAX_VALUE, Double.MAX_VALUE);

        // Setup the context menu
        final ContextMenu contextMenu = new ContextMenu();
        List<MenuItem> items = new ArrayList<>();
        getContextMenu(items);
        contextMenu.getItems().addAll(items);
        setContextMenu(contextMenu);

        // Set the drag and drop
        setOnDragDetected(new EventHandler<MouseEvent>() {
            @Override
            public void handle(MouseEvent event) {
                if (associatedComponent != null) {
                    Dragboard db = startDragAndDrop(TransferMode.MOVE);

                    SnapshotParameters sp = new SnapshotParameters();
                    sp.setFill(Color.TRANSPARENT);
                    Image snapshot = snapshot(sp, null);

                    double offsetX = 0;
                    double offsetY = 0;

                    if (OSValidator.isWindows()) {
                        offsetX = snapshot.getWidth()/2;
                        offsetY = snapshot.getHeight()/2;
                    }

                    db.setDragView(snapshot, offsetX, offsetY);

                    ClipboardContent content = new ClipboardContent();
                    content.putString(DragButton.DRAG_PREFIX+associatedComponent.json().toString());
                    db.setContent(content);

                    event.consume();
                }
            }
        });
        setOnDragDone(new EventHandler<DragEvent>() {
            public void handle(DragEvent event) {
                if (event.getTransferMode() == TransferMode.MOVE) {
                    if (onComponentActionListener != null) {
                        onComponentActionListener.onComponentDroppedAway();
                    }
                }

                setCursor(Cursor.DEFAULT);

                event.consume();
            }
        });
    }

    /**
     * Populate the context menu
     * @param items
     */
    protected void getContextMenu(List<MenuItem> items) {
        MenuItem edit = new MenuItem("Edit");
        edit.setOnAction(new EventHandler<ActionEvent>() {
            @Override
            public void handle(ActionEvent event) {
                if (onComponentActionListener != null) {
                    onComponentActionListener.onComponentEdit();
                }
            }
        });
        Image editImage = new Image(ComponentButton.class.getResourceAsStream("/assets/edit.png"));
        ImageView editImageView = new ImageView(editImage);
        editImageView.setFitWidth(16);
        editImageView.setFitHeight(16);
        edit.setGraphic(editImageView);

        MenuItem delete = new MenuItem("Delete");
        delete.setOnAction(new EventHandler<ActionEvent>() {
            @Override
            public void handle(ActionEvent event) {
                if (onComponentActionListener != null) {
                    onComponentActionListener.onComponentDelete();
                }
            }
        });
        Image deleteImage = new Image(ComponentButton.class.getResourceAsStream("/assets/delete.png"));
        ImageView deleteImageView = new ImageView(deleteImage);
        deleteImageView.setFitWidth(16);
        deleteImageView.setFitHeight(16);
        delete.setGraphic(deleteImageView);

        items.add(edit);
        items.add(delete);
    }

    protected String getParentTooltip() {
        if (associatedComponent.getItem().isValid()) {
            return "";
        }else{
            return "INVALID ITEM\n";
        }
    }

    public OnComponentActionListener getOnComponentActionListener() {
        return onComponentActionListener;
    }

    public void setOnComponentActionListener(OnComponentActionListener onComponentActionListener) {
        this.onComponentActionListener = onComponentActionListener;
    }

    public interface OnComponentActionListener {
        void onComponentEdit();
        void onComponentDelete();
        void onComponentDroppedAway();
    }
}

File: src/main/java/app/control_panel/layout_editor_tab/grid/button/ComponentButton.kt
package app.control_panel.layout_editor_tab.grid.button

import app.control_panel.layout_editor_tab.grid.GridContext
import app.control_panel.layout_editor_tab.grid.exception.CommandNotFoundException
import app.control_panel.layout_editor_tab.grid.dnd.ComponentDragReference
import javafx.event.EventHandler
import javafx.scene.CacheHint
import javafx.scene.Cursor
import javafx.scene.SnapshotParameters
import javafx.scene.control.ContentDisplay
import javafx.scene.control.ContextMenu
import javafx.scene.control.MenuItem
import javafx.scene.control.Tooltip
import javafx.scene.image.ImageView
import javafx.scene.input.ClipboardContent
import javafx.scene.input.TransferMode
import javafx.scene.paint.Color
import model.command.Command
import model.component.Component
import system.image.ImageResolver
import utils.OSValidator
import java.util.*


class ComponentButton(context : GridContext, val associatedComponent : Component) : SelectableButton(context) {
    // Used to request to the component grid which components are selected
    var requestSelectedComponentReference: (() -> ComponentDragReference)? = null

    var onComponentActionListener: OnComponentActionListener? = null

    init {
        // Get the associated command
        val command : Command? = context.commandResolver.getCommand(associatedComponent.commandId!!)

        if (command == null) {
            throw CommandNotFoundException()
        }

        // Set up the button image and text
        text = command.title

        // Create the tooltip
        command.description?.let {
            val tooltip = Tooltip()
            tooltip.text = it
            setTooltip(tooltip)
        }

        // Set up the image
        val imageView = ImageView()
        imageView.fitHeight = 48.0
        imageView.fitWidth = 48.0
        graphic = imageView
        contentDisplay = ContentDisplay.TOP

        context.imageResolver.loadInto(command.iconId, 48, imageView, fadeIn = false)

        // Add the style
        styleClass.add("component-btn")
        if (!command.isValid!!) {  // Invalid item rendering
            styleClass.add("invalid-btn")
        }

        // Setup the context menu
        val contextMenu = ContextMenu()
        val items = ArrayList<MenuItem>()
        getContextMenu(items)
        contextMenu.items.addAll(items)
        setContextMenu(contextMenu)

        // Set the drag and drop
        onDragDetected = EventHandler { event ->
            // Get the selected components
            var componentReference = requestSelectedComponentReference?.invoke()

            // If the dragged component is not in the selection, request to deselect all the others
            // and select only the current one. Then, request again the selected components
            if (associatedComponent !in componentReference?.components ?: emptyList()) {
                onDeselectAllRequest?.invoke()
                selected = true
                componentReference = requestSelectedComponentReference?.invoke()
            }

            if (componentReference != null) {


                // TODO: change image drag when there are multiple elements
                val db = startDragAndDrop(TransferMode.MOVE)

                val sp = SnapshotParameters()
                sp.fill = Color.TRANSPARENT
                val snapshot = snapshot(sp, null)

                var offsetX = 0.0
                var offsetY = 0.0

                if (OSValidator.isWindows()) {
                    offsetX = snapshot.width / 2
                    offsetY = snapshot.height / 2
                }

                db.setDragView(snapshot, offsetX, offsetY)

                val content = ClipboardContent()
                val payload = componentReference.json()
                content.putString(DragButton.Companion.COMPONENT_DRAG_PREFIX + payload.toString())
                db.setContent(content)
            }

            event.consume()
        }
        onDragDone = EventHandler { event ->
            if (event.transferMode == TransferMode.MOVE) {
                if (onComponentActionListener != null) {
                    onComponentActionListener!!.onComponentDroppedAway()
                }
            }

            cursor = Cursor.DEFAULT

            event.consume()
        }

        isCache = true
        cacheHint = CacheHint.SPEED
    }

    /**
     * Populate the context menu
     * @param items
     */
    protected fun getContextMenu(items: MutableList<MenuItem>) {
        val edit = MenuItem(context.resourceBundle.getString("edit"))
        edit.onAction = EventHandler {
            if (onComponentActionListener != null) {
                onComponentActionListener!!.onComponentEdit()
            }
        }

        val editImage = ImageResolver.getImage(ComponentButton::class.java.getResourceAsStream("/assets/edit.png"), 16)
        val editImageView = ImageView(editImage)
        editImageView.setFitWidth(16.0)
        editImageView.setFitHeight(16.0)
        edit.graphic = editImageView

        val delete = MenuItem(context.resourceBundle.getString("delete"))
        delete.onAction = EventHandler {
            if (onComponentActionListener != null) {
                onComponentActionListener!!.onComponentDelete()
            }
        }
        val deleteImage = ImageResolver.getImage(ComponentButton::class.java.getResourceAsStream("/assets/delete.png"), 16)
        val deleteImageView = ImageView(deleteImage)
        deleteImageView.setFitWidth(16.0)
        deleteImageView.setFitHeight(16.0)
        delete.graphic = deleteImageView

        items.add(edit)
        items.add(delete)
    }

    interface OnComponentActionListener {
        fun onComponentEdit()
        fun onComponentDelete()
        fun onComponentDroppedAway()
    }
}
